### ISSUE TITLE
Support queuing work from a worker thread (feature request)

### ISSUE THREAD

**Issue by mmomtchev** [libuv/libuv#3422]
This is a very significant evolution that can be accomplished with relatively minor modifications to the code.

`uv_queue_work()` should be thread-safe and support queuing work from any thread.

Queuing work from a worker thread will allow some very significant optimizations when chaining multiple independent operations that must run sequentially. Currently scheduling the next job requires synchronization with the main thread / event loop which in some cases is absolutely useless and leads to a very significant performance degradation.

Node.js `readFile` is a perfect example of this problem: <https://github.com/nodejs/node/issues/41435>

**Comment by mmomtchev** [comment]
I did some testing and found that:
* the performance of libuv in the new benchmark `uv_queue_work` matches very closely (less than 10% slower) the performance of `perf bench sched pipe` which is essentially the same test by the Linux kernel developers - a process waking up another process by using a pipe - without the additional overhead of libuv - meaning that it is pointless to try to optimize this further
* `eventfd` is only very marginally faster than a pipe on a modern Linux kernel - I have 17µs for a pipe vs 15µs for `eventfd` on my computer
* one can add a copious amount of additional mutexes without impacting performance - I clocked locking an uncontested mutex at 75ns on my computer - which is also the average cost of a *NOP* syscall

**Comment by mmomtchev** [comment]
@bnoordhuis @vtjnash As you can see from #3429, this can be retrofitted with no performance loss at all - the only change required is that the tracking of the number of pending I/O requests is to be atomic

However if this change is to be implemented, than it is also to be supported - so this means that queuing from the worker thread should not come with any hidden pitfalls - which, AFAIK, is indeed the case

Also, my PR reshuffles some of the header files in a significant way, do you have any opinion on this?

**Comment by realies** [comment]
@mmomtchev, I'm sorry to ask here, but do you know if the recent addition of C11 atomics (https://github.com/libuv/libuv/pull/3950) allowed for this, and subsequently https://github.com/nodejs/node/issues/41611 to go forward?

**Comment by mmomtchev** [comment]
@realies Not really, the groundwork is there - as now there are portable atomics in libuv - but the counting of the open file handles does not use them. The feature is only half way there.

**Comment by trevnorris** [comment]
FYI I started this work in https://github.com/libuv/libuv/pull/3938, but decided to split it up into smaller commits. Plan on coming back to this soon.
